### ISSUE TITLE
add `torchx list` command and `Runner.list` APIs

### ISSUE THREAD

**Issue by d4l3k** [pytorch/torchx#503]
## Description

Add a `torchx list` and `Runner/Scheduler.list` methods. This would allow listing all jobs the user has launched and see their status when tracking multiple different jobs. 

## Motivation/Background

Currently users have to use the scheduler specific tools like `sacct/vcctl/ray job list` to see all of their jobs. Adding this would allow users to just interact via the torchx interface and not have to worry about interacting with other tools.

## Detailed Proposal

We'd likely want something similar to https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.list

Filters may be hard to support across all schedulers so we probably want to limit it to just a few common ones or none at all initially. We also want to filter so we only return torchx jobs instead of all jobs on the scheduler.

Limiting it to jobs that the user owns would also be nice to have though may not be feasible for all schedulers.

## Additional context/links

* https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.list
* https://slurm.schedmd.com/sacct.html
* https://docs.aws.amazon.com/batch/latest/APIReference/API_ListJobs.html
* https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CustomObjectsApi.md#list_namespaced_custom_object
* https://docs.ray.io/en/master/cluster/jobs-package-ref.html#jobsubmissionclient

**Comment by anishchopra** [comment]
+1 for this, would love to have this feature

**Comment by d4l3k** [comment]
@anishchopra we'll likely add this in phases to support the various schedulers. What schedulers are you currently using just to prioritize and get early feedback?

We'd also be happy to get any other feedback you might have. Feel free to email me (tristanr@fb.com) or I can add you to the PT slack if you'd like to chat more

**Comment by anishchopra** [comment]
> @anishchopra we'll likely add this in phases to support the various schedulers. What schedulers are you currently using just to prioritize and get early feedback?
> 
> We'd also be happy to get any other feedback you might have. Feel free to email me ([tristanr@fb.com](mailto:tristanr@fb.com)) or I can add you to the PT slack if you'd like to chat more

I'm currently testing out the local_docker and kubernetes schedulers. Definitely happy to provide more feedback and join that slack channel. FYI, I'm an engineer at Rivian, and currently exploring all the different ways we can launch distributed training jobs in the cloud.

**Comment by priyaramani** [comment]
Hey @anishchopra 
Just merged our [first version](https://github.com/pytorch/torchx/pull/538) of the list API that lists torchx app handles launched on kubernetes volcano. You should be able to do `torchx list -s kuberentes` to list them. Please check it out and give us feedback. 
Also interested in learning about what features you would want the list command to support. 
My next steps looks like listing jobs started within a specific timeframe and supporting docker scheduler.

**Comment by yukunlin** [comment]
In the current API, the `list` method doesn't take any arguments. For the `aws_batch` scheduler, we'd need a way of at least specifying what queue to look at. Are the maintainers open to having it take arguments (like `scheduler_args` for the `run` command)?

**Comment by priyaramani** [comment]
> In the current API, the `list` method doesn't take any arguments. For the `aws_batch` scheduler, we'd need a way of at least specifying what queue to look at. Are the maintainers open to having it take arguments (like `scheduler_args` for the `run` command)?

Hi @yukunlin Thanks for asking! Yes we intend to support arguments to filter and list apps. Are there other filters that you think would be very helpful besides queue name?

**Comment by yukunlin** [comment]
Across all schedulers, I think support filtering by app ID name (maybe with regex too?) would be relevant. For the aws batch scheduler specifically, it would also be nice to be able to filter by a job's`shareIdentifier` (https://github.com/pytorch/torchx/pull/544).

**Comment by priyaramani** [comment]
Hey folks, the feature to list jobs and their statuses is now added! 

You can list jobs and their statuses using TorchX CLI or APIs without having to setup and use all the different scheduler specific tools. Right now torchx list defaults to listing jobs in all statuses. 

## CLI USAGE
`torchx list -s <scheduler_name>`
Example: 
![Screen Shot 2022-08-24 at 11 30 06 AM](https://user-images.githubusercontent.com/87679608/186502845-711638ee-db89-4461-befa-74992b535c68.png)

## PROGRAMMATIC USAGE
When you have a torchx runner object in python, just do
```
runner = torchx.runner.get_runner(...)
runner.list("scheduler_name")
```
#Example
`runner.list("kubernetes")`
#API spec
list(scheduler: str) → List[[torchx.schedulers.api.ListAppResponse](https://staticdocs.thefacebook.com/torchx/schedulers.html#torchx.schedulers.api.ListAppResponse)]
For apps launched on the scheduler, this API returns a list of ListAppResponse objects each of which have app id, app handle and its status.

**Comment by priyaramani** [comment]
@anishchopra @yukunlin 
Try it out and please keep the feedback coming!

**Comment by priyaramani** [comment]
@anishchopra @yukunlin Please try the feature and provide feedback at https://forms.gle/BgvgZFR78LmTsqm19 Thanks!
